addons/binding/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/api/HeosAPI.java

```diff
@@ -0,0 +1,315 @@
+     * Information of the sid and cid has to be obtained via the browse function
+     *
+     * @param pid The player ID where the media object shall be played
+     * @param sid The source ID where the media is located
+     * @param cid The container ID of the media
+     */
+    public void addContainerToQueuePlayNow(String pid, String sid, String cid) {
+        controller.send(controller.command().addContainerToQueuePlayNow(pid, sid, cid));
+    }
+
+    /**
+     * Sets the connection parameter if the HOES system and connects to the system
+     *
+     * @param ip The IP address of the HEOS player which is used as bridge
+     * @param port The port the system shall establish the connection
+     */
+    public void setHeosConnection(String ip, int port) {
+        controller.setConnectionIP(ip);
+        controller.setConnectionPort(port);
+        controller.establishConnection(false);
+    }
+
+    /**
+     * Reboot the bridge to which the connection is established
+     */
+    public void reboot() {
+        controller.sendWithoutResponse(controller.command().rebootSystem());
+    }
+
+    /**
+     * Login in via the bridge to the HEOS account
+     *
+     * @param name The username
+     * @param password The password of the user
+     */
+    public void logIn(String name, String password) {
+        controller.command().setUsernamePwassword(name, password);
+        controller.send(controller.command().signIn(name, password));
+    }
+
+    /**
+     * Plays a specific station on the HEOS player
+     *
+     * @param pid The player ID
+     * @param sid The source ID where the media is located
+     * @param cid The container ID of the media
+     * @param mid The media ID of the media
+     * @param name Station name returned by 'browse' command.
+     */
+    public void playStation(String pid, String sid, String cid, String mid, String name) {
+        controller.send(controller.command().playStation(pid, sid, cid, mid, name));
+    }
+
+    /**
+     * Plays a specified input source on the player.
+     * Set {@code source_pid} to null if destination and source is the same
+     * Input name as per specified in HEOS CLI Protocol
+     *
+     * @param des_pid the PID where the source shall be played
+     * @param source_pid the PID where the source is located. NULL if destination and source are the same
+     * @param input the input name
+     */
+    public void playInputSource(String des_pid, String source_pid, String input) {
+        if (source_pid == null) {
+            controller.send(controller.command().playInputSource(des_pid, des_pid, input));
+        } else {
+            controller.send(controller.command().playInputSource(des_pid, source_pid, input));
+        }
+    }
+
+    /**
+     * Plays a file from a URL
+     *
+     * @param pid the PID where the file shall be played
+     * @param url the complete URL the file is located
+     */
+    public void playURL(String pid, String url) {

```

Comment:
Good idea. Unfortunately the URL information is already delivered as string by the AudioStream class. I think a check if it is a proper URL is already done their. 
For other inputs checking for proper URL should maybe done somewhere else.
I'm also not so common with URL checking methods. If you have an idea please let me know.

Reply:
As this is a kind of separate component, it would still be better to request a URL instead of a String here, even if it requires two lines of extra code (maybe one day you would want to carve out the Heos API part from the binding and offer it as a library for generic usage also in other software projects). Syntax checking is automatically done if you use the built-in `java.net.URL` class of java. If the URL is malformed, a `MalformedURLException` will be thrown already at the object instantiation.